File: solutions/ruby/sieve/1/sieve.rb
class Sieve
  def initialize(limit)
    @limit = limit
  end

  def primes
    return [] if @limit < 2
    sieve = [true] * (@limit + 1)
    sieve[0] = false
    sieve[1] = false
    (2..Math.sqrt(@limit)).each do |i|
      if sieve[i]
        (i * i..@limit).step(i) { |j| sieve[j] = false }
      end
    end
    sieve.each_index.select { |i| sieve[i] }
  end
end

File: solutions/javascript/sieve/1/sieve.js
export const primes = (n) => {
  if (n < 2) return [];
  const sieve = new Array(n + 1).fill(true);
  sieve[0] = sieve[1] = false;
  for (let i = 2; i * i <= n; i++) {
    if (sieve[i]) {
      for (let j = i * i; j <= n; j += i) {
        sieve[j] = false;
      }
    }
  }
  return sieve.reduce((acc, isPrime, i) => {
    if (isPrime) acc.push(i);
    return acc;
  }, []);
};
